File: code/rustdb-core/src/storage/page.rs
pub const PAGE_SIZE: usize = 16384;
pub const MAGIC: u32 = 0x52444200; // "RDB\0"
pub const VERSION: u16 = 1;

#[derive(Debug, Clone)]
pub struct PageHeader {
    pub magic: u32,
    pub version: u16,
    pub row_count: u32,
    pub page_count: u32,
}

impl PageHeader {
    pub fn new() -> Self {
        PageHeader {
            magic: MAGIC,
            version: VERSION,
            row_count: 0,
            page_count: 0,
        }
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        let mut buf = Vec::new();
        buf.extend_from_slice(&self.magic.to_le_bytes());
        buf.extend_from_slice(&self.version.to_le_bytes());
        buf.extend_from_slice(&[0u8; 2]); // padding
        buf.extend_from_slice(&self.row_count.to_le_bytes());
        buf.extend_from_slice(&self.page_count.to_le_bytes());
        buf.extend_from_slice(&[0u8; 16]); // reserved
        buf // 32 bytes
    }

    pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
        if bytes.len() < 32 { return None; }
        let magic = u32::from_le_bytes(bytes[0..4].try_into().ok()?);
        if magic != MAGIC { return None; }
        let version = u16::from_le_bytes(bytes[4..6].try_into().ok()?);
        let row_count = u32::from_le_bytes(bytes[8..12].try_into().ok()?);
        let page_count = u32::from_le_bytes(bytes[12..16].try_into().ok()?);
        Some(PageHeader { magic, version, row_count, page_count })
    }
}

File: code/backend/core/src/storage/page.cpp
#include "engine/storage/page.hpp"

#include <cstring>

namespace engine {

namespace {
void push_le32(std::vector<std::uint8_t>& buf, std::uint32_t v) {
    for (int i = 0; i < 4; i++) buf.push_back(static_cast<std::uint8_t>((v >> (8 * i)) & 0xFF));
}
void push_le16(std::vector<std::uint8_t>& buf, std::uint16_t v) {
    for (int i = 0; i < 2; i++) buf.push_back(static_cast<std::uint8_t>((v >> (8 * i)) & 0xFF));
}
std::uint32_t read_le32(const std::uint8_t* p) {
    return static_cast<std::uint32_t>(p[0]) | (static_cast<std::uint32_t>(p[1]) << 8) |
           (static_cast<std::uint32_t>(p[2]) << 16) | (static_cast<std::uint32_t>(p[3]) << 24);
}
std::uint16_t read_le16(const std::uint8_t* p) {
    return static_cast<std::uint16_t>(p[0]) | static_cast<std::uint16_t>(p[1] << 8);
}
} // namespace

std::vector<std::uint8_t> PageHeader::to_bytes() const {
    std::vector<std::uint8_t> buf;
    buf.reserve(32);
    push_le32(buf, magic);   // 0..4
    push_le16(buf, version); // 4..6
    buf.push_back(flags);   // 6
    buf.push_back(0);       // 7 (padding)
    push_le32(buf, row_count);  // 8..12
    push_le32(buf, page_count); // 12..16
    buf.resize(32, 0);          // 16..32 reserved
    return buf; // 32 bytes
}

std::optional<PageHeader> PageHeader::from_bytes(const std::uint8_t* bytes, std::size_t len) {
    if (len < 32) return std::nullopt;
    PageHeader h;
    h.magic = read_le32(bytes);
    if (h.magic != MAGIC) return std::nullopt;
    h.version = read_le16(bytes + 4);
    h.flags = bytes[6];
    h.row_count = read_le32(bytes + 8);
    h.page_count = read_le32(bytes + 12);
    return h;
}

} // namespace engine
